### PR TITLE
more on multi model calbacks

### PR DESCRIPTION
**Pull request by vlimant**
do not merge yet. there are still things to be done now on the optimizer side of things

### CODE REVIEW

**Comment by vlimant** [comment]
@duanders if you have comments

File: mpi_learn/mpi/process.py [hunk 1]
```diff
@@ -194,7 +191,16 @@ def recv(self, obj=None, tag=MPI.ANY_TAG, source=None, buffer=False, status=None
             source = self.parent_rank 
         tag_num = self.lookup_mpi_tag(tag)
         if buffer:
-            comm.Recv( obj, source=source, tag=tag_num, status=status )
+            if type(obj) == list:
```
Comment (duanders) [inline mpi_learn/mpi/process.py]: This looks hard to maintain because the Send logic needs to be identical to the Recv logic.  Can we template this so Send and Recv both run the same set of loops?

Comment (vlimant) [reply mpi_learn/mpi/process.py]: I tried to put the send and receive on the same footing indeed. I'll try the templating next.

**Comment by vlimant** [comment]
at this point, the training gets all the way through.

File: mpi_learn/train/model.py [hunk 2]
```diff
@@ -234,6 +234,14 @@ def save(self, *args,**kwargs):
             for m in self.models:
                 m.save( *args, **kwargs )
 
+class GANModel(MPIModel):
```
Comment (vlimant) [inline mpi_learn/train/model.py]: that's the hook for the gan model @svalleco

**Comment by vlimant** [comment]
@svalleco this is it ! 
python models/get_3d.py 
to create a well formatted input (I think the ones could even be removed)

then
mpirun -np 8 ./MPIDriver.py bs.json train_3d.list test_3d.list --loss dummy --epochs 3 --master-gpu --features-name X --labels-name y --tf
or 
mpirun -np 8 ./MPIDriver.py bs.json train_3d.list test_3d.list --loss dummy --epochs 3 --master-gpu --features-name X --labels-name y

up until it tries to run validate() which fails on test_on_batch not being implemented, yet

**Comment by svalleco** [comment]
that's great news!
I'll try to run and  add the test_on_batch.

**Comment by vlimant** [comment]
now with test_on_batch and refactored train_on_batch.
the mpi installation is screwed up because of the previous tests with horovod, and I cannot try across nodes ; I'll fix that, hopefully :-(

**Comment by vlimant** [comment]
fixed for the theano backend. issue is, now, that with tf, the get_weights is hanging (@duanders  ...) the same way we observed otherwise

**Comment by vlimant** [comment]
mpirun  -n 5   ./MPIGDriver.py bs5.json train_3d.list test_1_3d.list --loss dummy --epochs 3 --master-gpu --features-name X --labels-name y --tf --optimizer sgd
for training with mpi

python simple_train.py
for the standalone training using the same exact ingredients

so far it looks good numerically

**Comment by vlimant** [comment]
couple of changes for pretty printing and two networks, one with fixed batchnorm, one without

File: mpi_learn/mpi/manager.py [hunk 3]
```diff
@@ -83,7 +83,39 @@ def get_gpu_list(mem_lim = 5000):
         #print ("Found",dev,"free")
         return dev
                                                                                         
+            
+class MPIKFoldManager(MPIManager):
```
Comment (vlimant) [inline mpi_learn/mpi/manager.py]: hey doctor @duanders ! what do you think of this ?
